src/Spice86/ViewModels/DisassemblyViewModel.cs

```diff
@@ -18,327 +19,384 @@ namespace Spice86.ViewModels;
+
+            _logger.Debug("Instructions loaded, now contains {DebuggerLinesCount} instructions", DebuggerLines.Count);
 
-    [RelayCommand(CanExecute = nameof(IsPaused))]
-    private async Task GoToFunction(object? parameter) {
-        if (parameter is FunctionInfo functionInfo) {
-            await GoToAddress(functionInfo.Address.Linear);
+            // Verify that the current instruction is now in the collection
+            if (!TryGetLineByAddress(address, out debuggerLine)) {
+                _logger.Error("Address {Address} still not found after loading instructions", address);
+
+                throw new InvalidOperationException($"Current address {address} still not found in DebuggerLines after update");
+            }
+        } finally {
+            IsLoading = false;
         }
-    }
 
-    private async Task GoToAddress(uint address) {
-        StartAddress = ConvertUtils.ToHex32(address);
-        await UpdateDisassembly();
-        SelectedInstruction = Instructions.FirstOrDefault();
+        return debuggerLine;
     }
 
-    private bool UpdateDisassemblyCommandCanExecute() {
-        return IsPaused &&
-            !ScanForValidationErrors(nameof(StartAddress));
-    }
+    /// <summary>
+    ///     Updates the highlighting for the current instruction based on the current CPU state.
+    /// </summary>
+    private void UpdateCpuInstructionHighlighting() {
+        // Ensure we're on the UI thread
+        if (!Dispatcher.UIThread.CheckAccess()) {
+            Dispatcher.UIThread.Post(UpdateCpuInstructionHighlighting);
 
-    [ObservableProperty]
-    private bool _isLoading;
+            return;
+        }
 
-    [RelayCommand(CanExecute = nameof(UpdateDisassemblyCommandCanExecute))]
-    private async Task UpdateDisassembly() {
-        if (!TryParseAddressString(StartAddress, _state, out uint? startAddress)) {
+        // Skip if we're already updating the highlighting to prevent recursive calls
+        if (_isUpdatingHighlighting) {
             return;
         }
-        Instructions.Clear();
-        IsLoading = true;
-        Instructions.AddRange(await Task.Run(
-            () => DecodeCurrentWindowOfInstructions(startAddress.Value)));
-        SelectedInstruction = Instructions.FirstOrDefault();
-        UpdateHeader(SelectedInstruction?.Address);
-        IsLoading = false;
-    }
 
-    private List<CpuInstructionInfo> DecodeCurrentWindowOfInstructions(uint startAddress) {
-        return
-            _instructionsDecoder.DecodeInstructions(
-                startAddress,
-                NumberOfInstructionsShown);
-    }
+        _isUpdatingHighlighting = true;
 
-    private CpuInstructionInfo? _selectedInstruction;
+        try {
+            // Log the current state for debugging
+            _logger.Debug("Updating highlighting: CPU instruction address={CpuInstructionAddress}, Previous={PreviousInstructionAddress}", State.IpSegmentedAddress, _previousInstructionAddress);
```

Comment:
check log level first or remove this _logger.Debug line ?

Reply:
Done.